### PR TITLE
Add mock registration client

### PR DESCRIPTION
**Pull request by ivankelly**
For testing the client without having to boot zookeeper.

### CODE REVIEW

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/discover/MockRegistrationClient.java [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+import org.apache.bookkeeper.versioning.Versioned;
+
+/**
+ * Mock implementation of registration client.
+ */
+public class MockRegistrationClient implements RegistrationClient {
+    /**
+     * Listener to receive changes from the registration service.
+     */
+    /*interface RegistrationListener {
```
Comment (eolivelli) [inline bookkeeper-server/src/test/java/org/apache/bookkeeper/discover/MockRegistrationClient.java]: Nit: this is to be removed

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/discover/MockRegistrationClient.java [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+        executor.submit(() -> {
+                readOnlyBookieWatchers.add(listener);
+                promise.complete(null);
+            });
+        return promise;
+    }
+
+    @Override
+    public void unwatchReadOnlyBookies(RegistrationListener listener) {
+        executor.submit(() -> {
```
Comment (eolivelli) [inline bookkeeper-server/src/test/java/org/apache/bookkeeper/discover/MockRegistrationClient.java]: Should we block until completion?

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/discover/MockRegistrationClient.java [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import org.apache.bookkeeper.net.BookieSocketAddress;
+import org.apache.bookkeeper.versioning.LongVersion;
+import org.apache.bookkeeper.versioning.Versioned;
+
+/**
+ * Mock implementation of registration client.
```
Comment (eolivelli) [inline bookkeeper-server/src/test/java/org/apache/bookkeeper/discover/MockRegistrationClient.java]: What about adding a note that this implementation runs all of the callbacks in a single thread, to mimic how zk works?

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/discover/MockRegistrationClient.java (hunk 1 above)
Comment (ivankelly) [reply bookkeeper-server/src/test/java/org/apache/bookkeeper/discover/MockRegistrationClient.java]: Removed.

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/discover/MockRegistrationClient.java (hunk 2 above)
Comment (ivankelly) [reply bookkeeper-server/src/test/java/org/apache/bookkeeper/discover/MockRegistrationClient.java]: the interface indicates that we don't need to. This isn't used anywhere in bk, though dlog does use it in one place. for now I would prefer to leave it as is, and if in future the interface changes to allow waiting, change this too.

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/discover/MockRegistrationClient.java (hunk 3 above)
Comment (ivankelly) [reply bookkeeper-server/src/test/java/org/apache/bookkeeper/discover/MockRegistrationClient.java]: Added a comment.

**Comment by eolivelli** [comment]
retest this please

**Pull request closed without merging**